### PR TITLE
[doc] Prompty doc

### PR DESCRIPTION
**Pull request by lalala123123**
# Description
image

Add three doc of prompty:
- index
- prompty output format
- using prompty in flex flow

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: docs/how-to-guides/develop-a-prompty/index.md [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+model:
+    api: chat
+    configuration:
+      type: azure_openai
+      azure_deployment: gpt-35-turbo
+      connection: azure_open_ai_connection
+    parameters:
+      max_tokens: 128
+      temperature: 0.2
+inputs:
```
Comment (wangchao1230) [inline docs/how-to-guides/develop-a-prompty/index.md:40]: Shall we omit inputs outputs for now? As we are still finalizing the spec?
Or we should update after that?

File: docs/how-to-guides/develop-a-prompty/index.md [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+:::
+
+Promptflow provides the `prompty` feature to provide customers with a simple way to develop prompt template.
+
+## Create a prompty
+In promptflow, a file with a `.prompty` extension will be treated as a prompty.
+
+### Prompty specification
+The prompty asset is a markdown file with a modified front matter. 
+The front matter is in yaml format that contains a number of metadata fields which defines model configuration and expected inputs of the prompty.
```
Comment (wangchao1230) [inline docs/how-to-guides/develop-a-prompty/index.md:14]: explain the content is a jinja tempalate?

File: docs/how-to-guides/develop-a-prompty/index.md [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+from promptflow.core import Prompty
+
+# Load prompty with dict override
+override_model = {
+    "configuration": {
+        "api_key": "${env:AZURE_OPENAI_API_KEY}",
+        "api_version": "${env:AZURE_OPENAI_API_VERSION}",
+        "azure_endpoint": "${env:AZURE_OPENAI_ENDPOINT}"
+    },
+    "parameters": {"max_token": 512}
```
Comment (wangchao1230) [inline docs/how-to-guides/develop-a-prompty/index.md:128]: max_tokens

File: docs/how-to-guides/develop-a-prompty/index.md [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+# Load prompty with dict override
+configuration = AzureOpenAIModelConfiguration(
+    azure_deployment="gpt-3.5-turbo",
+    api_key="${env:AZURE_OPENAI_API_KEY}",
+    api_version="${env:AZURE_OPENAI_API_VERSION}",
+    azure_endpoint="${env:AZURE_OPENAI_ENDPOINT}"
+)
+override_model = {
+    "configuration": configuration,
+    "parameters": {"max_token": 512}
```
Comment (wangchao1230) [inline docs/how-to-guides/develop-a-prompty/index.md:147]: max_tokens

File: docs/how-to-guides/develop-a-prompty/index.md [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+
+:::{tab-item} SDK
+:sync: SDK
+```python
+from promptflow.client import PFClient
+
+pf = PFClient() 
+# create run
+prompty_run = pf.run(
+    flow=r"C:\project\promptflow\src\promptflow\tests\test_configs\prompty\prompty_example.prompty",
```
Comment (wangchao1230) [inline docs/how-to-guides/develop-a-prompty/index.md:369]: avoid such long path

File: docs/how-to-guides/develop-a-prompty/prompty-output-format.md [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+```text
+Ah, the age-old question about the meaning of life! 🌍🤔 The meaning of life is a deeply philosophical and subjective topic. Different people have different perspectives on it. Some believe that the meaning of life is to seek happiness and fulfillment, while others find meaning in personal relationships, accomplishments, or spiritual beliefs. Ultimately, it's up to each individual to explore and discover their own purpose and meaning in life. 🌟
+```
+
+### Json object output
+When the user meets the following conditions, prompty returns content of first choices as a dict.
+- Define response_format to type: json_object in parameters
+- Specify the return json format in template.
+
+**Note**: response_format is compatible with GPT-4 Turbo and all GPT-3.5 Turbo models newer than gpt-3.5-turbo-1106. For more details, refer to this [document](https://platform.openai.com/docs/api-reference/chat/create#chat-create-response_format).
```
Comment (wangchao1230) [inline docs/how-to-guides/develop-a-prompty/prompty-output-format.md:60]: add \` \` on model name like: `GPT-4 Turbo`

File: docs/how-to-guides/develop-a-prompty/index.md [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+```python
+from promptflow.core import Prompty
+
+prompty_obj = Prompty.load(source="path/to/prompty.prompty")
+result = prompty_obj(first_name="John", last_name="Doh", question="What is the capital of France?")
+```
+
+### Test prompty
+
+#### Test as a flow
```
Comment (wangchao1230) [inline docs/how-to-guides/develop-a-prompty/index.md:250]: may add some screenshot on prompty trace

File: docs/how-to-guides/index.md [hunk 8]
```diff
@@ -17,6 +17,14 @@ chat-with-a-flow/index
 run-and-evaluate-a-flow/index
 ```
 
+```{toctree}
+:caption: Prompty
+:maxdepth: 1
+develop-a-prompty/index
+develop-a-prompty/prompty-output-format
```
Comment (wangchao1230) [inline docs/how-to-guides/index.md:24]: no need to have these two as index covers them?

File: docs/how-to-guides/develop-a-prompty/index.md [hunk 9]
```diff
@@ -0,0 +1,384 @@
+# Prompty
+
+:::{admonition} Experimental feature
+This is an experimental feature, and may change at any time. Learn [more](../faq.md#stable-vs-experimental).
+:::
+
+Promptflow provides the `prompty` feature to provide customers with a simple way to develop prompt template.
```
Comment (brynn-code) [inline docs/how-to-guides/develop-a-prompty/index.md:7]: 2 'provide' here, maybe asking copilot for some suggestions

File: docs/how-to-guides/develop-a-prompty/index.md (hunk 1 above)
Comment (lalala123123) [reply docs/how-to-guides/develop-a-prompty/index.md]: remove inputs/output in docs

File: docs/how-to-guides/develop-a-prompty/index.md (hunk 2 above)
Comment (lalala123123) [reply docs/how-to-guides/develop-a-prompty/index.md]: add explain that jinja format

File: docs/how-to-guides/develop-a-prompty/index.md (hunk 3 above)
Comment (lalala123123) [reply docs/how-to-guides/develop-a-prompty/index.md]: fix it

File: docs/how-to-guides/develop-a-prompty/prompty-output-format.md (hunk 6 above)
Comment (lalala123123) [reply docs/how-to-guides/develop-a-prompty/prompty-output-format.md]: fix it

File: docs/how-to-guides/develop-a-prompty/index.md (hunk 9 above)
Comment (zhengfeiwang) [reply docs/how-to-guides/develop-a-prompty/index.md:7]: +1, maybe replace 2nd provide to enable

File: docs/how-to-guides/index.md (hunk 8 above)
Comment (lalala123123) [reply docs/how-to-guides/index.md:24]: I think it's better to keep these two doc. The prompt structure and execution method are briefly introduced in index.
The other two documents introduce the more complex configuration methods of prompt and the usage of nesting into other types of flows.

File: docs/how-to-guides/develop-a-prompty/index.md (hunk 7 above)
Comment (lalala123123) [reply docs/how-to-guides/develop-a-prompty/index.md:236]: add screenshot about prompty trace

File: docs/how-to-guides/develop-a-prompty/index.md [hunk 10]
```diff
@@ -0,0 +1,374 @@
...
+
+## Create a prompty
+In promptflow, a file with a `.prompty` extension will be treated as a prompty.
+
+### Prompty specification
+The prompty asset is a markdown file with a modified front matter.
```
Comment (zhengfeiwang) [inline docs/how-to-guides/develop-a-prompty/index.md:13]: is "front matter" a term in prompty? not quite understand its meaning...

File: docs/how-to-guides/develop-a-prompty/index.md [hunk 11]
```diff
@@ -0,0 +1,374 @@
...
+```
+
+:::
+::::
+
+## Execute a prompty
+
+Prompty can be executed in these ways to meet the needs of customers in different scenarios.
+
+### Prompty as a call
```
Comment (zhengfeiwang) [inline docs/how-to-guides/develop-a-prompty/index.md:225]: will "as a function" better?

File: docs/how-to-guides/develop-a-prompty/index.md [hunk 12]
```diff
@@ -0,0 +1,374 @@
...
+# Test prompty with default inputs
+pf flow test --flow path/to/prompty.prompty
+
+# Test prompty with specified inputs
+pf flow test --flow path/to/prompty.prompty --inputs first_name=John last_name=Doh question="What is the capital of France?"
+
+# Test prompty with sample file
+pf flow test --flow path/to/prompty.prompty --inputs path/to/sample.json
+```
+There will be a trace link shown in terminal to visualize the internal execution details for this run.
```
Comment (zhengfeiwang) [inline docs/how-to-guides/develop-a-prompty/index.md:252]:
```suggestion
There will be a trace link shown in terminal to visualize the internal execution details for this command.
```

Comment (zhengfeiwang) [reply docs/how-to-guides/develop-a-prompty/index.md:252]: not sure run is proper here, as above command is a flow test

File: docs/how-to-guides/develop-a-prompty/prompty-output-format.md [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+# Customer
+You are helping {{first_name}} {{last_name}} to find answers to their questions.
+Use their name to address them in your responses.
+
+user:
+{{question}}
+```
+
+### Streaming output
+When `stream=true` is configured in the parameters of a prompty whose output format is text, promptflow sdk will return a generator type, which item is the content of each chunk.
```
Comment (zhengfeiwang) [inline docs/how-to-guides/develop-a-prompty/prompty-output-format.md:187]: for the last sentence, might be "which yields each chunk"? - maybe let GPT help on this.

File: docs/how-to-guides/develop-a-prompty/index.md (hunk 9 above)
Comment (lalala123123) [reply docs/how-to-guides/develop-a-prompty/index.md]: copilot refines it to "Promptflow introduces the `prompty` feature, it is designed to simplify the development of prompt templates for customers."

File: docs/how-to-guides/develop-a-prompty/index.md (hunk 11 above)
Comment (lalala123123) [reply docs/how-to-guides/develop-a-prompty/index.md]: refine to "Direct function call"

File: docs/how-to-guides/develop-a-prompty/prompty-output-format.md (hunk 13 above)
Comment (lalala123123) [reply docs/how-to-guides/develop-a-prompty/prompty-output-format.md]: refine it by copilot, "Each item from the generator represents the content of a chunk."

File: docs/how-to-guides/develop-a-prompty/index.md (hunk 12 above)
Comment (lalala123123) [reply docs/how-to-guides/develop-a-prompty/index.md]: refine it

**Review by wangchao1230 — DISMISSED** [review]

**Review by wangchao1230 — DISMISSED** [review]

**Review by Adrian999999999 — APPROVED** [review]

File: docs/how-to-guides/develop-a-prompty/index.md [hunk 14]
```diff
@@ -0,0 +1,392 @@
...
+After this front matter is the prompt template, articulated in the `Jinja` format.
+
+Fields in the front matter:
+
+| Field       | Description                                                                                              |
+|-------------|----------------------------------------------------------------------------------------------------------|
+| name        | The name of the prompt.                                                                                  |
+| description | A description of the prompt.                                                                             |
+| model       | Details the prompty's model configuration, including connection info and parameters for the LLM request. |
+| sample      | Offers a dictionary or JSON file containing sample data for inputs and outputs.                          |
```
Comment (wangchao1230) [inline docs/how-to-guides/develop-a-prompty/index.md:25]: add back inputs outpus in next PR

File: docs/how-to-guides/develop-a-prompty/index.md [hunk 15]
```diff
@@ -0,0 +1,392 @@
...
+model:
+  api: chat
+  configuration:
+    type: azure_openai
+    azure_deployment: gpt-35-turbo
+    connection: azure_open_ai_connection
+  parameters:
+    max_tokens: 128
+    temperature: 0.2
+sample:
```
Comment (wangchao1230) [inline docs/how-to-guides/develop-a-prompty/index.md:40]: update this part

File: docs/how-to-guides/develop-a-prompty/index.md [hunk 16]
```diff
@@ -0,0 +1,392 @@
...
+  configuration:
+    type: azure_openai
+    azure_deployment: gpt-35-turbo
+    api_key: <api-key>
+    api_version: <api-version>
+    azure_endpoint: <azure-endpoint>
+  parameters:
+    max_tokens: 128
+    temperature: 0.2
+sample:
```
Comment (wangchao1230) [inline docs/how-to-guides/develop-a-prompty/index.md:84]: update

File: docs/how-to-guides/develop-a-prompty/index.md [hunk 17]
```diff
@@ -0,0 +1,392 @@
...
+  configuration:
+    type: openai
+    model: gpt-3.5-turbo
+    api_key: <api-key>
+    base_url: <api_base>
+  parameters:
+    max_tokens: 128
+    temperature: 0.2
+sample:
+  inputs:
```
Comment (wangchao1230) [inline docs/how-to-guides/develop-a-prompty/index.md:163]: update

File: docs/how-to-guides/develop-a-prompty/index.md [hunk 18]
```diff
@@ -0,0 +1,392 @@
...
+  configuration:
+    type: azure_open_ai
+    azure_deployment: gpt-35-turbo
+    connection: azure_open_ai_connection
+
+  parameters:
+    max_tokens: 128
+    temperature: 0.2
+sample:
+  inputs:
```
Comment (wangchao1230) [inline docs/how-to-guides/develop-a-prompty/index.md:304]: udpate

File: docs/how-to-guides/develop-a-prompty/index.md (hunk 14 above)
Comment (lalala123123) [reply docs/how-to-guides/develop-a-prompty/index.md]: fix it

File: docs/how-to-guides/develop-a-prompty/index.md (hunk 15 above)
Comment (lalala123123) [reply docs/how-to-guides/develop-a-prompty/index.md:51]: update it

File: docs/how-to-guides/develop-a-prompty/index.md [hunk 19]
```diff
@@ -37,11 +39,19 @@ model:
   parameters:
     max_tokens: 128
     temperature: 0.2
+inputs:
+  first_name:
+    type: string
+    default: John
```
Comment (wangchao1230) [inline docs/how-to-guides/develop-a-prompty/index.md:45]: remove default, let's only discuss this in our future doc describe signature part

Comment (lalala123123) [reply docs/how-to-guides/develop-a-prompty/index.md]: remove default

**Review by wangchao1230 — APPROVED** [review]

**Review by wangchao1230 — APPROVED** [review]

**Review by crazygao — APPROVED** [review]

**Pull request merged**